File: 01-uebungsaufgaben/musterloesungen/fachwerk-python/truss.py
import numpy as np
import matplotlib.pyplot as plt

from matplotlib.patches import Polygon
from matplotlib.collections import PatchCollection


def truss_p(EA, x1, x2, u1, u2):
    l0 = np.linalg.norm(x2 - x1)
    l = np.linalg.norm(x2 + u2 - (x1+u1))
    d = 1 / l * (x2 + u2 - (x1+u1))
    t = np.concatenate((-d, d))
    return EA * (l - l0) / l0 * t


def truss_ke(EA, x1, x2):
    l0 = np.linalg.norm(x2 - x1)
    d0 = (x2 - x1) / l0
    t0 = np.concatenate((-d0, d0))
    return EA / l0 * np.outer(t0, t0)


def truss_n(EA, x1, x2, u1, u2):
    l0 = np.linalg.norm(x2 - x1)
    d0 = (x2 - x1) / l0
    return EA / l0 * np.dot(u2 - u1, d0)


def truss_p_lin(EA, x1, x2, u1, u2):
    K = truss_ke(EA, x1, x2)
    u = np.concatenate((u1, u2))
    return K @ u


def draw_system(s):
    
    # Knoten
    plt.scatter(s.nodes[0, :], s.nodes[1, :], color='black', zorder=2, s=15)
    
    # Elemente
    for e in range(s.elements.shape[1]):
        n = s.elements[:, e]
        plt.plot(s.nodes[0, n], s.nodes[1, n], color = 'black')    
    
    # Kräfte
    plt.quiver(
        s.nodes[0, :], s.nodes[1, :], 
        s.forces[0, :], s.forces[1, :], 
        color = 'red', width=0.003
    )
    
    # Auflager
    sn = s.supports[0, :]
    sx = s.nodes[0, sn]
    sy = s.nodes[1, sn]
    plt.scatter(sx, sy, marker=5, s=150, color='tab:blue')
    sn = s.supports[1, :]
    sx = s.nodes[0, sn]
    sy = s.nodes[1, sn]
    plt.scatter(sx, sy, marker=6, s=150, color='tab:blue')
    
    # Anzeigen
    plt.axis('equal')
    plt.show()


def solve(s):

    # Anzahl Knoten und Freiheitsgrade
    Nn = s.nodes.shape[1]
    Ne = s.elements.shape[1]
    N = 2 * Nn
    
    # Globale Steifigkeitsmatrix
    K = np.zeros([N, N])
    for e in range(0, Ne):
        n1, n2 = s.elements[:, e]
        x1 = s.nodes[:, n1]
        x2 = s.nodes[:, n2]
        Ke = truss_ke(s.EA, x1, x2)
        I = [2*n1, 2*n1+1, 2*n2, 2*n2+1]    
        K[np.ix_(I, I)] += Ke

    # Lastvektor
    F = np.reshape(s.forces, N, order='F')

    # Auflager
    idx = np.reshape(s.supports, N, order='F')
    F[idx] = 0
    K[idx, :] = 0
    K[idx, idx] = 1

    # LGS lösen und Verschiebungen in Matrix ablegen
    s.u_hat = np.reshape(np.linalg.solve(K, F), (2, Nn), order='F')
    s.N = member_forces(s)


def member_forces(s):
    Ne = s.elements.shape[1]
    N = np.zeros(Ne)
    
    for e in range(0, Ne):
        n1, n2 = s.elements[:, e]
        x1 = s.nodes[:, n1]
        x2 = s.nodes[:, n2]
        u1 = s.u_hat[:, n1]
        u2 = s.u_hat[:, n2]
        N[e] = truss_n(s.EA, x1, x2, u1, u2)
    
    return N


def draw_deformation(s, scale):
    
    # Knotenkoordinaten verschoben
    nu = s.nodes + scale * s.u_hat
    
    # Knoten
    plt.scatter(nu[0, :], nu[1, :], color='black', zorder=2, s=15)
    
    # Elemente
    for e in range(s.elements.shape[1]):
        n = s.elements[:, e]
        plt.plot(s.nodes[0, n], s.nodes[1, n], color = 'gray', zorder=-5)
        plt.plot(nu[0, n], nu[1, n], color = 'black')    
    
    # Auflager
    sn = s.supports[0, :]
    sx = nu[0, sn]
    sy = nu[1, sn]
    plt.scatter(sx, sy, marker=5, s=150, color='tab:blue')
    sn = s.supports[1, :]
    sx = nu[0, sn]
    sy = nu[1, sn]
    plt.scatter(sx, sy, marker=6, s=150, color='tab:blue')
    
    # Anzeigen
    plt.axis('equal')
    plt.show()


def draw_member_forces(s, scale):
    
    # Liste von Patches    
    patches = []
    
    # Patches für Normalkraft und Elemente plotten
    for e in range(s.elements.shape[1]):
        n1, n2 = s.elements[:, e]
        x1 = s.nodes[:, n1]
        x2 = s.nodes[:, n2]

        e1 = np.array(x2 - x1) / np.linalg.norm(x2 - x1)
        normal = np.array([e1[1], -e1[0]])
        o = scale *s.N[e] * normal
        patches.append(Polygon(np.row_stack((x1, x2, x2 + o, x1 + o))))
        
        plt.plot([x1[0], x2[0]], [x1[1], x2[1]], color = 'black')    
    
    # Betragsmaessig groesste Normalkraft
    max_n = np.max(np.abs(s.N))

    # Plotten
    pc = PatchCollection(patches, edgecolor='black', alpha=0.6)
    pc.set_array(s.N)
    pc.set_clim(-max_n, max_n)
    pc.set_cmap('bwr')
    plt.gca().add_collection(pc)
    plt.axis('equal')
    plt.colorbar(pc)
    plt.show()

File: 01-uebungsaufgaben/musterloesungen/fachwerk-julia/truss.jl
function truss_p(EA, x1, x2, u1, u2)
    l0 = norm(x2 - x1)
    l = norm(x2 + u2 - (x1 + u1))
    d = 1 / l * (x2 + u2 - (x1 + u1))
    t = vcat(-d, d)
    return EA * (l - l0) / l0 * t
end


function truss_k(EA, x1, x2)
    l0 = norm(x2 - x1)
    d0 = (x2 - x1) / l0
    t0 = vcat(-d0, d0)
    return EA / l0 * t0 * t0'
end


function truss_p_lin(EA, x1, x2, u1, u2)
    K = truss_k(EA, x1, x2)
    u = vcat(u1, u2)
    return K * u
end


function truss_n(EA, x1, x2, u1, u2)
    l0 = norm(x2 - x1)
    d0 = (x2 - x1) / l0
    return EA / l0 * (u2 - u1) ⋅ d0
end


function struct_size(s)
    mm1 = extrema(s.nodes[1, :])
    l1 = mm1[2] - mm1[1]
    mm2 = extrema(s.nodes[2, :])
    l2 = mm2[2] - mm1[1]
    return sqrt(l1 * l2)
end


function draw_system(s)
    sz = struct_size(s)
    Nn = size(s.nodes, 2)

    # Graphik und Achsen
    fig = Figure()
    Axis(fig[1, 1], aspect=DataAspect())

    # Elemente
    linesegments!(reshape(s.nodes[:, s.elements], 2, :), color=:black, linewidth=3)

    # Auflager
    h = 1 / sqrt(3)
    s1 = 0.05 * sz * [0 -1 -1 0; 0 -h h 0]
    s2 = 0.05 * sz * [0 -h h 0; 0 -1 -1 0]
    for i = 1:Nn
        x = s.nodes[:, i]
        s.supports[1, i] && lines!(x .+ s1, color=:black)
        s.supports[2, i] && lines!(x .+ s2, color=:black)
    end

    # Kräfte
    f = 0.2 * sz / maximum(norm.(eachcol(s.forces))) * s.forces
    arrows!(s.nodes[1, :], s.nodes[2, :], f[1, :], f[2, :], color=:red)

    # Knoten
    scatter!(s.nodes, color=:white, strokewidth=1)

    # Grafik zurückgeben
    return fig
end


function member_forces(s)
    Ne = size(s.elements, 2)
    N = zeros(Ne)
    for e = 1:Ne
        n1, n2 = s.elements[:, e]
        x1 = s.nodes[:, n1]
        x2 = s.nodes[:, n2]
        u1 = s.u_hat[:, n1]
        u2 = s.u_hat[:, n2]
        N[e] = truss_n(s.EA, x1, x2, u1, u2)
    end
    return N
end


function solve!(s)
    Nn = size(s.nodes, 2)
    Ne = size(s.elements, 2)
    N = 2 * Nn

    # Steifigkeitsmatrix
    K = zeros(N, N)
    for e = 1:Ne
        n1, n2 = s.elements[:, e]
        x1 = s.nodes[:, n1]
        x2 = s.nodes[:, n2]
        Ke = truss_k(s.EA, x1, x2)
        I = [2 * n1 - 1, 2 * n1, 2 * n2 - 1, 2 * n2]
        K[I, I] += Ke
    end

    # Lastvektor
    F = copy(reshape(s.forces, N))

    # Auflager
    idx = reshape(s.supports, N)
    F[idx] .= 0
    K[idx, :] .= 0
    K[diagind(K)[idx]] .= 1

    # Verschiebungen und Normalkraft
    s.u_hat = reshape(K \ F, 2, Nn)
    s.N = member_forces(s)
end


function draw_deformations(s)
    sz = struct_size(s)
    Nn = size(s.nodes, 2)
    max_u = maximum(norm.(eachcol(s.u_hat)))
    xu = s.nodes + 0.1 * sz / max_u * s.u_hat

    # Graphik und Achsen
    fig = Figure()
    Axis(fig[1, 1], aspect=DataAspect())

    # Elemente
    linesegments!(reshape(s.nodes[:, s.elements], 2, :), color=:gray)
    linesegments!(reshape(xu[:, s.elements], 2, :), linewidth=3, color=:black)

    # Auflager
    h = 1 / sqrt(3)
    s1 = 0.05 * sz * [0 -1 -1 0; 0 -h h 0]
    s2 = 0.05 * sz * [0 -h h 0; 0 -1 -1 0]
    for i = 1:Nn
        x = xu[:, i]
        s.supports[1, i] && lines!(x .+ s1, color=:black)
        s.supports[2, i] && lines!(x .+ s2, color=:black)
    end

    # Kräfte
    max_f = maximum(norm.(eachcol(s.forces)))
    f = 0.2 * sz / max_f * s.forces
    arrows!(xu[1, :], xu[2, :], f[1, :], f[2, :], color=:red)

    # Knoten
    scatter!(xu, color=:white, strokewidth=1)

    # Grafik zurückgeben
    return fig
end


function draw_member_forces(s)
    fig = Figure()
    Axis(fig[1, 1], aspect=DataAspect())

    # Elemente
    for n in eachcol(s.elements)
        lines!(s.nodes[:, n], color=:black, linewidth=3)
    end

    # Polygone sammeln
    polys = []
    sz = struct_size(s)
    Ne = size(s.elements, 2)
    max_n = maximum(abs.(s.N))
    scale = 0.1 * sz / max_n
    for e = 1:Ne
        x1 = s.nodes[:, s.elements[1, e]]
        x2 = s.nodes[:, s.elements[2, e]]
        e1 = normalize(x2 - x1)
        normal = [e1[2], -e1[1]]
        offset = scale * s.N[e] * normal
        push!(polys, Point2f[x1, x2, x2+offset, x1+offset])
    end

    # Polygone plotten
    p = poly!(
        polys, color=s.N, strokewidth=1, alpha=0.5,
        colormap=:redblue, colorrange=[-max_n, max_n]
    )
    Colorbar(fig[1, 2], p)

    # Grafik zurückgeben
    return fig
end